main/src/main/scala/sbt/Defaults.scala

```diff
@@ -1569,6 +1574,46 @@ object Classpaths {
     val x2 = copyResources.value
     classDirectory.value :: Nil
   }
+  // This is a variant of exportProductsTask with tracking
+  private[sbt] def trackedExportedProducts(track: TrackLevel): Initialize[Task[Classpath]] = Def.task {
+    val art = (artifact in packageBin).value
+    val module = projectID.value
+    val config = configuration.value
+    for { (f, analysis) <- trackedProductsImplTask(track).value } yield APIMappings.store(analyzed(f, analysis), apiURL.value).put(artifact.key, art).put(moduleID.key, module).put(configuration.key, config)
+  }
+  private[this] def trackedProductsImplTask(track: TrackLevel): Initialize[Task[Seq[(File, Analysis)]]] =
+    Def.taskDyn {
+      val useJars = exportJars.value
+      val jar = (artifactPath in packageBin).value
+      val dirs = productDirectories.value
+      TrackLevel.intersection(track, exportToInternal.value) match {
+        case TrackLevel.TrackAlways if (useJars) =>
+          Def.task {
+            Seq((packageBin.value, compile.value))
+          }
+        case TrackLevel.TrackAlways =>
+          Def.task {
+            products.value map { (_, compile.value) }
+          }
+        case TrackLevel.TrackIfMissing if (useJars && !jar.exists) =>
+          Def.task {
+            Seq((packageBin.value, compile.value))
+          }
+        case TrackLevel.TrackIfMissing if (!useJars && (dirs flatMap { d => (d ** "*.class").get }).nonEmpty) =>
```

Comment:
Added 02c72392040cb3aef57fb49e7fa8e5864ced18c8. Interestingly, there's a slight performance penalty in terms of time when I break the iteration apart into directories and then `*.class`. Adds around 25% more time when tested with `new File(".")`, traversing all directories under sbt projects. For more realistic test traversing `new File("util/io/target/classes")`, the penalty remained: 3ms vs 2ms using the original implementation. *Hopefully*, it's doing better in memory perf, but I'm not too sure about that either without profiling.